gems/sorbet-runtime/lib/types/props/optional.rb

```diff
@@ -46,8 +51,30 @@ def compute_derived_rules(rules)
     rules[:need_nil_read_check] = T::Props::Utils.need_nil_read_check?(rules)
   end
 
+  # checked(:never) - O(runtime object construction)
+  sig {returns(T::Hash[Symbol, T::Props::Private::ApplyDefault]).checked(:never)}
+  attr_reader :props_with_defaults
```

Comment:
having this as hash of arbitrary code executions make me cringe. 
Is it possible to make this code more explicit and not pass arbitrary "can do anything" functions?
E.g. descripe what it indents to do with interface/datastructure that's more specific than `Proc`?

Reply:
In this case `ApplyDefault` is a struct, not a proc. Are you thinking of the `SetterProc` thing below?

Reply:
Yes, sorry :-)

Reply:
I'm struggling a bit to find a solution that improves this. The issue is something like this:
1. We want to share the code we'd use for a setter, so that validation logic is guaranteed to be the same;
2. We don't always have a setter (`const` properties don't have one) so we can't just use `send`;
3. The setter code is used as a method body, so it assumes `self` is the T::Props instance, which implies it needs to be run with `instance_exec` if it hasn't actually been bound as a method;
4. I don't want to add a layer of indirection between what `props_without_defaults` gives us and what we can pass to `instance_exec`, to avoid a runtime cost.

We could have Setter be a subclass of Proc, but that doesn't really seem any better to me than just using a type alias. I'm fine with making that change if you prefer it, though. Or do you have any other ideas?